resolver-dns/src/main/java/io/netty/resolver/dns/UnixResolverDnsServerAddressStreamProvider.java

```diff
@@ -0,0 +1,233 @@
+     */
+    public UnixResolverDnsServerAddressStreamProvider(String etcResolvConf, String etcResolverDir) throws IOException {
+        this(etcResolvConf == null ? null : new File(etcResolvConf),
+             etcResolverDir == null ? null :new File(etcResolverDir).listFiles());
+    }
+
+    @Override
+    public DnsServerAddressStream nameServerAddressStream(String hostname) {
+        for (;;) {
+            int i = hostname.indexOf('.', 1);
+            if (i < 0 || i == hostname.length() - 1) {
+                return defaultNameServerAddresses != null ? defaultNameServerAddresses.stream() : null;
+            }
+
+            DnsServerAddresses addresses = domainToNameServerStreamMap.get(hostname);
+            if (addresses != null) {
+                return addresses.stream();
+            }
+
+            hostname = hostname.substring(i + 1);
+        }
+    }
+
+    boolean mayOverrideNameServers() {
+        return !domainToNameServerStreamMap.isEmpty() ||
+                defaultNameServerAddresses != null && defaultNameServerAddresses.stream().next() != null;
+    }
+
+    private static Map<String, DnsServerAddresses> parse(File... etcResolverFiles) throws IOException {
+        Map<String, DnsServerAddresses> domainToNameServerStreamMap =
+                new HashMap<String, DnsServerAddresses>(etcResolverFiles.length << 1);
+        for (File etcResolverFile : etcResolverFiles) {
+            if (!etcResolverFile.isFile()) {
+                continue;
+            }
+            FileReader fr = new FileReader(etcResolverFile);
+            BufferedReader br = null;
+            try {
+                br = new BufferedReader(fr);
+                List<InetSocketAddress> addresses = new ArrayList<InetSocketAddress>(2);
+                String domainName = etcResolverFile.getName();
+                int port = DNS_PORT;
+                String line;
+                while ((line = br.readLine()) != null) {
+                    line = line.trim();
+                    char c;
+                    if (line.isEmpty() || (c = line.charAt(0)) == '#' || c == ';') {
+                        continue;
+                    }
+                    if (line.startsWith(NAMESERVER_ROW_LABEL)) {
+                        int i = indexOfNonWhiteSpace(line, NAMESERVER_ROW_LABEL.length());
+                        if (i < 0) {
+                            throw new IllegalArgumentException("error parsing label " + NAMESERVER_ROW_LABEL +
+                                    " in file " + etcResolverFile + ". value: " + line);
+                        }
+                        String maybeIP = line.substring(i);
+                        // There may be a port appended onto the IP address so we attempt to extract it.
+                        if (!NetUtil.isValidIpV4Address(maybeIP) && !NetUtil.isValidIpV6Address(maybeIP)) {
+                            i = maybeIP.lastIndexOf('.');
+                            if (i + 1 >= maybeIP.length()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS134)